### PR TITLE
Add MoreGraaskamp strategy

### LINKED ISSUE #1051: Specific strategy request: Graaskamp
This strategy is one of the strategies from Axelrod's original tournaments. It is described here http://axelrod.readthedocs.io/en/stable/reference/overview_of_strategies.html#graaskamp as:

1. Play Tit For Tat for the first 50 rounds;
2. Defects on round 51;
3. Plays 5 further rounds of Tit For Tat;
4. A check is then made to see if the opponent is playing randomly in which case it defects for the rest of the game;
5. The strategy also checks to see if the opponent is playing Tit For Tat or another strategy from a preliminary tournament called ‘Analogy’. If so it plays Tit For Tat. If not it cooperates and randomly defects every 5 to 15 moves.

Step 4 can be implemented using a chi squared test similarly to the Stein and Rapaport strategy which is implemented here: https://github.com/Axelrod-Python/Axelrod/blob/master/axelrod/strategies/axelrod_first.py#L496

Step 5 is a bit tricky as it refers to a strategy called "Analogy". I have not been able to find any information describing this strategy. I suggest that anyone who tries to work on this does their best to find that information but if there is non available let us simply ignore that part of the check **but clearly point that out in the documentation of the strategy**.

### PR DESCRIPTION
**Pull request by SplitInfinity**
This pull request adds an implementation of the Graaskamp strategy from
Axelrod's second tournament.

Resolves #1051

### CODE REVIEW

File: axelrod/strategies/graaskamp.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    Names:
+
+    - GRASR: [Axelrod1993]_
+    """
+
+    # These are various properties of the strategy
+    name = 'MoreGraaskamp'
+    classifier = {
+        'memory_depth': 5,
```
Comment (marcharper) [inline axelrod/strategies/graaskamp.py]: The `memory_depth` is `float('inf')` (because of the use of the total score)

File: axelrod/strategies/graaskamp.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        self.defect_round = 0
+
+    # The strategy itself
+    def strategy(self, opponent: Player) -> Action:
+        round_number = len(self.history) + 1
+
+        # Tit for tat for first 50 rounds
+        if round_number == 1:
+            return C
+        elif round_number < 51:
```
Comment (marcharper) [inline axelrod/strategies/graaskamp.py]: These `elif` can all be `if` because of the returns

File: axelrod/strategies/graaskamp.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                if round_number == 118:
+                    self.mode = Mode.TitForTat
+
+                return C
+            else:
+                # self.mode must be Mode.DefectRandomly
+                if round_number == self.defect_round:
+                    self.defect_round = round_number + randrange(5, 15) + 1
+                    return D
+                else:
```
Comment (marcharper) [inline axelrod/strategies/graaskamp.py]: This `else` is unnecessary

File: axelrod/tests/strategies/test_graaskamp.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+from .test_player import TestPlayer
+
+C, D = axelrod.Action.C, axelrod.Action.D
+
+class TestMoreGraaskamp(TestPlayer):
+
+    name = 'MoreGraaskamp'
+    player = axelrod.MoreGraaskamp
+    expected_classifier = {
+        'memory_depth': 5,
```
Comment (marcharper) [inline axelrod/tests/strategies/test_graaskamp.py]: `float('inf')`

File: axelrod/strategies/graaskamp.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
+from axelrod.action import Action, actions_to_str
+from axelrod.player import Player
+from axelrod.random_ import randrange
+from axelrod._strategy_utils import calculate_scores
+
+from enum import Enum
+
+C, D = Action.C, Action.D
+
+class Mode(Enum):
```
Comment (drvinceknight) [inline axelrod/strategies/graaskamp.py]: Can we remove this class and just use a string as the corresponding attribute in the player class.

File: axelrod/strategies/graaskamp.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        1 - If none of the below modes are selected, cooperate and defect randomly
+            every 5 to 15 turns
+        2 - If the opponent's score is greater than 135 and moves 52 through 55 were
+            [C, D, C, D] or [D, D, D, D], play tit for tat thereafter starting with
+            cooperate
+        3 - If the opponent's score is less than or equal to 135, always defect 
+            hereafter
+        4 - If the opponent's score is greater than 135 and moves 52 through 56 were
+            [C, C, D, C, C], cooperate until and including move 118, and tit for tat
+            thereafter
```
Comment (drvinceknight) [inline axelrod/strategies/graaskamp.py]: We would then need to move the mode descriptions to here.

File: axelrod/strategies/graaskamp.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.mode = Mode.Unknown
```
Comment (drvinceknight) [inline axelrod/strategies/graaskamp.py]: `self.mode = "unknown"`

File: axelrod/strategies/graaskamp.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        # Play 5 further rounds of tit for tat
+        elif round_number < 57:
+            return opponent.history[-1]
+        elif round_number == 57:
+            game = self.match_attributes['game']
+
+            self_score, opponent_score = calculate_scores(self, opponent, game)
+
+            if opponent_score <= 135:
+                self.mode = Mode.AlwaysDefect
```
Comment (drvinceknight) [inline axelrod/strategies/graaskamp.py]: `self.mode = "always_defect"`

Comment (drvinceknight) [inline axelrod/strategies/graaskamp.py]: Same for other modes.

File: axelrod/strategies/graaskamp.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            cooperate
+        3 - If the opponent's score is less than or equal to 135, always defect 
+            hereafter
+        4 - If the opponent's score is greater than 135 and moves 52 through 56 were
+            [C, C, D, C, C], cooperate until and including move 118, and tit for tat
+            thereafter
+
+    Names:
+
+    - GRASR: [Axelrod1993]_
```
Comment (drvinceknight) [inline axelrod/strategies/graaskamp.py]: Can you also add:

```
- Graaskamp: [Axelrod1980b]_
```

File: axelrod/tests/strategies/test_graaskamp.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
+"""Tests for the Graaskamp strategies."""
+
+import axelrod
+from .test_player import TestPlayer
+
+C, D = axelrod.Action.C, axelrod.Action.D
+
+class TestMoreGraaskamp(TestPlayer):
```
Comment (drvinceknight) [inline axelrod/tests/strategies/test_graaskamp.py]: These tests look good but can you add an `attrs` check (for each) which verifies the strategy is in the mode it should be. An example here: https://github.com/Axelrod-Python/Axelrod/blob/master/axelrod/tests/strategies/test_oncebitten.py#L38

File: axelrod/strategies/graaskamp.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        
+class MoreGraaskamp(Player):
+    """
+    This strategy can be described as follows:
+    
+    1. Play tit for tat for the first 50 rounds.
+    2. Defect on round 51.
+    3. Play 5 further rounds of tit for tat.
+    4. In round 57, select between four different "modes" for the rest of the game:
+        1 - If none of the below modes are selected, cooperate and defect randomly
```
Comment (meatballs) [inline axelrod/strategies/graaskamp.py]: typo: should be 'is selected' rather than 'are selected'

**Comment by drvinceknight** [comment]
> Thanks for this, I have tested it locally and it agrees with the implemented strategy.

Actually my test didn't pick this strategy up because it's stochastic. Once you've made the requested changes we will check the behaviour as best as possible to ensure it matches up. One way to do this would be to confirm the fingerprints are the same. Here's the Fortran code's fingerprint: https://github.com/Axelrod-Python/Axelrod-fingerprint#grasr

The code to run this would be:

```
>>> import axelrod as axl
>>> fp = axl.AshlockFingerprint(axl.MoreGrasskamp)
>>> fp.fingerprint(turns=200, repetitions=100)
>>> fp.plot();
```

They should give similar images. If they don't that indicates there's something not quite right.

**Comment by marcharper** [comment]
Was there something wrong with the enum?

**Comment by SplitInfinity** [comment]
@marcharper 

I don't think so. @drvinceknight requested that I replace it with a simple string member that records the mode, so that's what I did.

**Comment by drvinceknight** [comment]
> Was there something wrong with the enum?

I didn't see what advantage it offered to just a string for the mode (which is akin to what is done elsewhere). Less code is good code :)

Thanks for all the modifications @SplitInfinity, I'm going to run the fingerprint locally and will post it here :+1:

**Comment by drvinceknight** [comment]
It looks like the logic must be a bit off. Here is the fingerprint of the fortran code:



Here is a reproduced one (over 20 repetitions):

download 2

**Comment by SplitInfinity** [comment]
@drvinceknight Do the visual differences between the fingerprints give any hints as to where the differences in logic are?

**Comment by drvinceknight** [comment]
> @drvinceknight Do the visual differences between the fingerprints give any hints as to where the differences in logic are?

Ish... These fingerprints aren't great if I'm honest (not their implementation but the theory behind them), they're implemented directly from the literature.

So the area where they seem to disagree is for x >.5 and x>y>.25 (ish). In that area: `grasr` seem to be getting a higher score than `MoreGrassman`. This would indicate that in that area `grasr` is defecting more than `MoreGrassman`.

The fingerprints are calculated by pitting the strategies `grasr` and `MoreGrassman` against a given strategy for every (x, y) with the following definition (http://axelrod.readthedocs.io/en/stable/tutorials/further_topics/fingerprinting.html):

- x of the time the strategy cooperates
- y of the time the strategy defects
- 1-x-y of the time the strategy plays Tit For Tat

There is an interpretation for when x + y >= 1 which (it looks like is where our problem region is):

- 1 - x of the time the strategy cooperates
- 1 - y of the time the strategy defects
- 1 - x - y of the time the strategy plays the opposite of Tit For Tat

**TLDR**

I THINK (but these fingerprints are hard to interpret) that when the opponent defects a lot (because it's mainly playing a defector (1-x is low) or the opposite of TfT) `MoreGrassman` is not defecting enough.

**Comment by drvinceknight** [comment]
Another (better) clue:

```python
>>> for opponent in axl.basic_strategies:
...     turns = 500
...     matches = [axl.Match((grassr, opponent()), turns=turns),
...                axl.Match((moreG, opponent()), turns=turns)]
...     for match in matches:
...         axl.seed(0)
...         match.play()
...     print([match.final_score_per_turn()[0] for match in matches], opponent.name)
[2.492, 1.726] Alternator
[4.692, 4.692] Anti Tit For Tat
[4.684, 4.684] Bully
[3.176, 3.156] Cooperator
[2.548, 2.502] Cycler DC
[0.998, 0.998] Defector
[1.204, 1.204] Grudger
[1.15, 1.15] Suspicious Tit For Tat
[2.994, 2.918] Tit For Tat
[2.526, 1.752] Win-Shift Lose-Stay
[2.164, 1.812] Win-Stay Lose-Shift
```

Seems to suggest that there is a discrepancy against `Win-Shift Lose-Stay`.

Confirmation:

```python
>>> import numpy as np
>>> repetitions = 50
>>> match = axl.Match((grassr, axl.WinShiftLoseStay()), turns=turns)
>>> scores = []
>>> for seed in range(repetitions):
...     axl.seed(seed)
...     match.play()
...     scores.append(match.final_score_per_turn()[0])
>>> np.mean(scores)
2.5259999999999998
``` 

```python
>>> match = axl.Match((moreG, axl.WinShiftLoseStay()), turns=turns)
>>> scores = []
>>> for seed in range(repetitions):
...     axl.seed(seed)
...     match.play()
...     scores.append(match.final_score_per_turn()[0])
>>> np.mean(scores)
1.7615599999999998
```

For reference here is the source code for `WinShiftLoseStay`: http://axelrod.readthedocs.io/en/stable/_modules/axelrod/strategies/memoryone.html#WinShiftLoseStay. It's not terribly well described there but:

1. It starts by defecting;
2. If it cooperates and the opponent defects it cooperates in the next round
3. If it defects and the opponent cooperates it cooperates in the next round
4. If both cooperate it defects
5. If both defect it defects

**Comment by drvinceknight** [comment]
Here's the confirmation against `Alternator`:

```python
>>> match = axl.Match((grassr, axl.Alternator()), turns=turns)
>>> scores = []
>>> for seed in range(repetitions):
...     axl.seed(seed)
...     match.play()
...     scores.append(match.final_score_per_turn()[0])
>>> np.mean(scores)
2.4920000000000004
```

```python
>>> match = axl.Match((moreG, axl.Alternator()), turns=turns)
>>> scores = []
>>> for seed in range(repetitions):
...     axl.seed(seed)
...     match.play()
...     scores.append(match.final_score_per_turn()[0])
>>> np.mean(scores)
1.7347600000000003
```

**Comment by drvinceknight** [comment]
The fact that the scores seem to be relatively similar against `CyclerDC` which is essentially `Alternator` but with a different initial action (starts with `D`) is probably a big clue:

```python
>>> match = axl.Match((grassr, axl.CyclerDC()), turns=turns)
>>> scores = []
>>> for seed in range(repetitions):
...     axl.seed(seed)
...     match.play()
...     scores.append(match.final_score_per_turn()[0])
>>> np.mean(scores)
2.5445199999999999
```

```python
>>> match = axl.Match((moreG, axl.CyclerDC()), turns=turns)
>>> scores = []
>>> for seed in range(repetitions):
...     axl.seed(seed)
...     match.play()
...     scores.append(match.final_score_per_turn()[0])
>>> np.mean(scores)
2.5019999999999993
```

**Comment by drvinceknight** [comment]
With 100 reps (I had it running, feel like I might as well throw it in here):

download 3

**Comment by SplitInfinity** [comment]
I'm trying to run these little snippets myself so I can change the number of turns and figure out when the scores start to differ, but I'm having some trouble using the `axelrod_fortran` module. This is what happens when I try to import it in the interpreter after installing it with `pip install axelrod-fortran`:

`OSError: libstrategies.so: cannot open shared object file: No such file or directory`

Are there additional setup steps I'm not aware of?

**Comment by meatballs** [comment]
@SplitInfinity We've not written much in the way of documentation for this, but you need to have compiled and installed the fortran code from https://github.com/Axelrod-Python/TourExec

The axelrod-fortran library is a Python wrapper for the compiled result.

Note - at present, this isn't available on Windows

**Comment by marcharper** [comment]
@SplitInfinity I can give you some sample outputs of the Fortran strategy for testing purposes, say every opponent history of length 10 and the response? Would that help?

**Comment by SplitInfinity** [comment]
@marcharper 

The first 55 or so turns of Graaskamp are essentially tit for tat so I don't expect the issue to be there. Round 57 and the ones after are more likely to reveal the issue, since the strategy enters one of the four modes at that point. To start with, could you send moves made by the Fortran strategy in rounds 1 - 100 against an Alternator? Thanks in advance!

**Comment by marcharper** [comment]
```
import axelrod as axl
import axelrod_fortran as axlf
player = axlf.Player("grasr")
opponent = axl.Alternator()
m = axl.Match((player, opponent), turns=100)
m.play()
print(m.result)
```
```
[(C, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (C, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D), (D, C), (C, D)]

**Comment by marcharper** [comment]
[This gist](https://gist.github.com/marcharper/eab8ad75c11f9da5e4e32004144a1040) has matches with many more opponents.

**Comment by drvinceknight** [comment]
Keep in mind that Graaskamp is stochastic, trying to match up to exact interactions is perhaps tricky but looking at the limiting scores in my snippets (over repetitions of matches) is a good indication. Although you can't reproduce the fortran ones locally at least you know what you're aiming for with `MoreGraaskamp`.

**Comment by marcharper** [comment]
I can fix the random seed if that will help

**Comment by drvinceknight** [comment]
> I can fix the random seed if that will help

It would do no harm but still wouldn't allow for a direct comparison.

The fortran strategies don't necessarily follow the same stochastic path (a random number is sampled every turns regardless of whether or not it is used for example). So really the scores are the best indication we can get.

**Comment by drvinceknight** [comment]
> I suggest using my matches above as an indication of what MoreGraaskamp should score.

This code snippet (and similar above) does not need `axelrod_fortran`:

```python
>>> import axelrod as axl
>>> moreG = axl.MoreGraaskamp()
>>> match = axl.Match((moreG, axl.CyclerDC()), turns=turns)
>>> scores = []
>>> for seed in range(repetitions):
...     axl.seed(seed)
...     match.play()
...     scores.append(match.final_score_per_turn()[0])
>>> np.mean(scores)
2.5019999999999993
```

**Comment by marcharper** [comment]
This one has been stale for a while -- is there a way we can make progress? Or merge (with a docstring that it's not exactly correct as is) and open an issue to figure out the mismatch?

**Comment by drvinceknight** [comment]
> is there a way we can make progress? 

Yes: one of us can take the time to go through the Fortran and the PR and offer guidance and help.

> Or merge (with a docstring that it's not exactly correct as is) and open an issue to figure out the mismatch?

No. We should not do that.

File: axelrod/strategies/graaskamp.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+                # Opponent's moves 52 through 55
+                prev_four_opponent_actions = opponent.history[-5:-1]
+
+                if last_five_opponent_actions == [C, C, D, C, C]:
+                    self.mode = 'cooperate_then_tit_for_tat'
+                elif prev_four_opponent_actions == [C, D, C, D] or prev_four_opponent_actions == [D, D, D, D]:
+                    self.mode = 'tit_for_tat'
+                else:
+                    self.mode = 'defect_randomly'
+                    self.defect_round = round_number + randrange(5, 15) + 1
```
Comment (marcharper) [inline axelrod/strategies/graaskamp.py]: This line may be slightly off: the Fortran code has `N = RANDO * 10.0 + 5.0`. If I compare `randrange(5, 15) + 1` with `5 + int(10 * random.random())` there seems to be an off by one shift.

Can we try that (also change the instance below), and also change to `defect_round` `rounds_until_defection` and count down by one each round rather than trying to compute what future round to defect on? I know it may be equivalent but it's closer to the original implementation, and a little easier to match them up.

File: axelrod/strategies/graaskamp.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+                return C
+
+        # Beyond round 57, obey the rules of the mode entered during round 57
+        if self.mode == 'tit_for_tat':
```
Comment (marcharper) [inline axelrod/strategies/graaskamp.py]: Can we add comments for the original enum here? I.e. this mode `IGAME = 2` for example.

File: axelrod/strategies/graaskamp.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+            if opponent_score <= 135:
+                self.mode = 'always_defect'
+                return D
+            else:
+                # Opponent's moves 52 through 56
+                last_five_opponent_actions = opponent.history[-5:]
+                # Opponent's moves 52 through 55
+                prev_four_opponent_actions = opponent.history[-5:-1]
+
+                if last_five_opponent_actions == [C, C, D, C, C]:
```
Comment (marcharper) [inline axelrod/strategies/graaskamp.py]: If the random issue doesn't fix everything my guess is that there is a transcription error in this line or the next few compared to the Fortran:

```
31    IF (NMOV(1) - 3) 12, 35, 12
35    IF (NMOV(3) - 3) 12, 16, 12
32    IF (NMOV(1) - 5) 12, 33, 12
33    IF (NMOV(3) - 5) 12, 16, 12
30    IF (NMOV(1) - 2) 12, 34, 12
34    IF (NMOV(3) - 4) 12, 40, 12
40 IF (NMOV(4) - 2) 12, 41, 12
```
In particular, it seems like there is at least one more way to get to the `round_number == 118` check (the jump to line 41 on the last line above) than is currently possible with the python implementation.

**Comment by drvinceknight** [comment]
@SplitInfinity just checking if you were going to work on this anymore? No problem if you've got other things on. I was thinking of finding a moment to work on this, I'd fork from your branch so when (eventually) merged in, all your commits would be there and it would auto merge this PR (which we can leave open until then). :) :+1:

**Comment by SplitInfinity** [comment]
@drvinceknight Probably not :(
Sorry guys!

**Comment by drvinceknight** [comment]
> Probably not :(
> Sorry guys!

No problem at all, I'll pick it up, thanks for your efforts on this! :)

**Comment by drvinceknight** [comment]
I doubt I'm going to find time to come around to this so am closing, if anyone would like to pick it up please do.
